torch/csrc/autograd/anomaly_mode.h

```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include "torch/csrc/utils/python_stub.h"
```

Comment:
Is there a way to make this code Python agnostic? It's nearly so, except for the store/print stack functionality. We're trying to expunge all of the `NO_PYTHON` bits from the codebase.

Reply:
That's closer.

The hidden constraint I'm surfacing here is that we have designs on splitting _C.so into two libraries: libtorch.so (no Python dependency) and libtorch_python.so (Python dependency). So having NO_PYTHON in the cpp file goes against this.

One usual way to go about doing this is to have anomaly_mode and anomaly_mode_python.

Reply:
Alternately, we can just merge this, and let the poor sap who actually ends up splitting this stuff fix it XD

Reply:
The reason why I save the traces as a python object is because the metadata is a python object anyway and I want to do minimal job at every Function creation and do the full formatting only if an actual error occurs.
The thing is that the metadata object associated with the Function itself is a python dictionnary (I think we want this to stay because from a user perspective, we want to be able to save any python object into a Function metadata field). So the same change will be needed for the function.h file to extend its abilities.

I am not that fluent in cpp, how would you implement that exactly? How can you make anomaly_mode_python modifying the original behavior if anomaly_mode? It cannot redefine the same function with different implementations right, can it change a flag?

Reply:
Change the metadata to be a class instead of a Python object. You also need to move some functionality from AnomalyMode to the new class.

Rough design:

Make AnomalyMode not use any Python. It just contains is_enabled/set_enabled.

```
// pure virtual interface
struct AnomalyMetadata {
  virtual void store_stack() = 0;
  virtual void print_stack() = 0;
};

// Python implementation
struct PyAnomalyMetadata : public AnomalyMetadata {
  PyAnomalyMetadata() { /* make dict */ }
  virtual ~PyAnomalyMetadata() { /* acquire GIL and release */ }
  virtual void store_stack();
  virtual void print_stack();

  PyObject* dict();
  
  PyObject* dict_;  // never null
};

struct Function {
  ...
  std::unique_ptr<AnomalyMetadata> metadata_;
};
```

Then you just need a factory to the correct concrete AnomalyMetadata instance. I would add it to Engine/PythonEngine.

```
struct Engine {
  virtual std::unique_ptr<AnomalyMetadata> make_anomaly_metadata();
}
struct PythonEngine {
  virtual std::unique_ptr<AnomalyMetadata> make_anomaly_metadata() override;
}
```

Reply:
That sounds good, thanks a lot for the help !